core/src/main/java/hudson/model/AsyncAperiodicWork.java

```diff
@@ -91,8 +141,54 @@ public void run() {
     }
 
     protected StreamTaskListener createListener() {
+        File f = getLogFile();
+        if (!f.getParentFile().isDirectory()) {
+            if (!f.getParentFile().mkdirs()) {
+                logger.log(getErrorLoggingLevel(), "Could not create directory {0}", f.getParentFile());
+            }
+        }
+        if (f.isFile()) {
+            if ((lastRotateMillis + logRotateMillis < System.currentTimeMillis())
+                    || (logRotateSize > 0 && f.length() > logRotateSize)) {
+                lastRotateMillis = System.currentTimeMillis();
+                File p = null;
+                for (int i = 5; i >= 0; i--) {
+                    File o = i == 0 ? f : new File(f.getParentFile(), f.getName() + "." + i);
+                    if (o.isFile()) {
+                        if (p != null && !p.exists()) {
+                            if (!o.renameTo(p)) {
+                                logger.log(getErrorLoggingLevel(), "Could not rotate log files {0} to {1}",
+                                        new Object[]{o, p});
+                            }
+                        } else {
+                            if (!o.delete()) {
+                                logger.log(getErrorLoggingLevel(), "Could not delete log file {0} to enable rotation",
+                                        o);
+                            }
+                        }
+                    }
+                    p = o;
```

Comment:
_nit_ the naming choice of these variables makes it a bit hard to follow along with what is supposed to happen.